drivers/gpio/gpio_pca95xx.c

```diff
@@ -0,0 +1,671 @@
+ *
+ * @param dev Device struct of the PCA95XX.
+ * @param reg Register to read (the PORT0 of the pair of registers).
+ * @param buf Buffer to read data into.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int read_port_regs(struct device *dev, u8_t reg,
+			   union gpio_pca95xx_port_data *buf)
+{
+	const struct gpio_pca95xx_config * const config =
+		dev->config->config_info;
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	struct device * const i2c_master = drv_data->i2c_master;
+	u16_t i2c_addr = config->i2c_slave_addr;
+	int ret;
+
+	ret = i2c_burst_read(i2c_master, i2c_addr, reg, buf->byte, 2);
+	if (ret) {
+		LOG_ERR("PCA95XX[0x%X]: error reading register 0x%X (%d)",
+			i2c_addr, reg, ret);
+		goto error;
+	}
+
+	LOG_DBG("PCA95XX[0x%X]: Read: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
+		i2c_addr, reg, buf->byte[0], (reg + 1), buf->byte[1]);
+
+error:
+	return ret;
+}
+
+/**
+ * @brief Write both port 0 and port 1 registers of certain register function.
+ *
+ * Given the register in reg, write the pair of port 0 and port 1.
+ *
+ * @param dev Device struct of the PCA95XX.
+ * @param reg Register to write into (the PORT0 of the pair of registers).
+ * @param buf Buffer to write data from.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int write_port_regs(struct device *dev, u8_t reg,
+			    union gpio_pca95xx_port_data *buf)
+{
+	const struct gpio_pca95xx_config * const config =
+		dev->config->config_info;
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	struct device * const i2c_master = drv_data->i2c_master;
+	u16_t i2c_addr = config->i2c_slave_addr;
+	int ret;
+
+	LOG_DBG("PCA95XX[0x%X]: Write: REG[0x%X] = 0x%X, REG[0x%X] = "
+		"0x%X", i2c_addr, reg, buf->byte[0], (reg + 1),
+		buf->byte[1]);
+
+	ret = i2c_reg_write_byte(i2c_master, i2c_addr, reg, buf->byte[0]);
+	if (ret) {
+		LOG_ERR("PCA95XX[0x%X]: error writing to register 0x%X "
+			"(%d)", i2c_addr, reg, ret);
+	}
+	ret = i2c_reg_write_byte(i2c_master, i2c_addr, reg+1, buf->byte[1]);
+	if (ret) {
+		LOG_ERR("PCA95XX[0x%X]: error writing to register 0x%X "
+				"(%d)", i2c_addr, reg, ret);
+	}
+	return ret;
+}
+
+/**
+ * @brief Setup the pin direction (input or output)
+ *
+ * @param dev Device struct of the PCA95XX
+ * @param access_op Access operation (pin or port)
+ * @param pin The pin number
+ * @param flags Flags of pin or port
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int setup_pin_dir(struct device *dev, int access_op,
+			  u32_t pin, int flags)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data *reg_dir = &drv_data->reg_cache.dir;
+	union gpio_pca95xx_port_data *reg_out = &drv_data->reg_cache.output;
+	int ret;
+
+	/* For each pin, 0 == output, 1 == input */
+	switch (access_op) {
```

Comment:
This switch is unnecessary since the function can only be reached when `access_op` is `GPIO_ACCESS_BY_PIN` or an invalid value: `GPIO_ACCESS_BY_PORT` is no longer passed into the driver API, and it's not passed internally in configuration operations.

Reply:
I keep these here because `gpio_config()` is still a valid public API, where `GPIO_ACCESS_BY_PORT` is still a valid `access_op`. These switch statements can be collapsed into simply `GPIO_ACCESS_BY_PIN` once the old APIs are removed.

Reply:
We would like to remove the `access_op` very soon. It would help us if the code was written in a way that makes it easy. Already today no API function will be called with `GPIO_ACCESS_BY_PORT`. It would be enough if at the beginning of `gpio_pca95xx_config` function you add the following code. Then all the dependencies can be cleaned up already.
```
	if (access_op != GPIO_ACCESS_BY_PIN) {
		return -ENOTSUP;
	}
```
I understand it's more work but it will make reviewing the code easier.

Reply:
No problem. I have uploaded a new version.